### ISSUE TITLE
S3 and Swift memory consumption

### ISSUE THREAD

**Issue by tnt-dev** [wal-g/wal-g#1253]
### Database name
PostgreSQL

# Issue description

### Describe your problem
wal-g consumes much more memory with S3 driver than with Swift.

Simple test, PostgreSQL 13 database ~20 GB:

```
/usr/bin/time -f "RSS=%M elapsed=%E cpu.sys=%S user=%U" envdir /etc/wal-g/env/ wal-g  backup-push -f /var/lib/postgresql/13/main
```

**Swift**:
```
RSS=128332 elapsed=6:27.24 cpu.sys=37.27 user=340.78
```
**S3**:

WALG_S3_MAX_PART_SIZE=52428800
WALG_UPLOAD_CONCURRENCY=1
```
RSS=838036 elapsed=8:08.81 cpu.sys=44.38 user=358.75
```

WALG_S3_MAX_PART_SIZE=52428800
WALG_UPLOAD_CONCURRENCY=16
```
RSS=3410616 elapsed=6:50.10 cpu.sys=48.49 user=355.12
```

WALG_S3_MAX_PART_SIZE=104857600
WALG_UPLOAD_CONCURRENCY=1
```
RSS=1851768 elapsed=6:39.94 cpu.sys=38.12 user=352.29
```

WALG_S3_MAX_PART_SIZE=104857600
WALG_UPLOAD_CONCURRENCY=16
```
RSS=3465728 elapsed=6:37.30 cpu.sys=48.38 user=356.49
```

So S3 driver consumes a lot of memory to upload backup in all combinations of part size and concurrency.
Swift works fast and doesn't need so much resources.

Do we have any ways to reduce wal-g memory consumption when use S3?

**Comment by x4m** [comment]
Hi! This seems very interesting, thanks for raising attention to this.
Can you please run benchmarks with disk throttling enabled? This is done with WALG_DISK_RATE_LIMIT in bytes per second. Storage adapter may have internal buffer that does not block incoming data and grows without restrictions...

**Comment by tnt-dev** [comment]
I profiled backup process and got that:
![profile001](https://user-images.githubusercontent.com/2064785/165498008-d3465174-8aeb-48a6-9d81-44f90717f1a0.png)
So problem is somewhere in AWS memory slice managment.
And looks like it was solved here: https://github.com/aws/aws-sdk-go/commit/3e8ec8863a560aa21f98ad480afa67df2e716859 and released in v1.34.30 

I have run my tests again with aws sdk v1.34.30 and got much better results!
pprof still shows that we have several extra memory allocations of WALG_S3_MAX_PART_SIZE and maybe that can be fixed in WAL-G but it requires additional debug... 

So the simplest solution for now is to bump aws/aws-sdk-go version. Which one should we use - v1.34.30 or current v1.44.1 is the only question.
